---
title: "README"
output: html_document
---
#R Assignment
##Part 1: File Inspection

To inspect files, I first read the files into R using the following: 
```{r Reading data}
fang <- read.table(file="working_files/fang_et_al_genotypes.txt", sep = "\t", header = TRUE, na.strings="?/?", stringsAsFactors = FALSE)
snps <- read.table(file="working_files/snp_position.txt",sep = "\t", header=TRUE, na.strings = "?/?", stringsAsFactors = FALSE )
```
After some fooling around in the help menu for read.table, I realized that the seperator needed to be specified because the files are not seperated by commas. The default behavior of reading strings as factors would cause issues as well, and I assigned the unknown values from the initial file (?/?) to be NAs using na.strings. 
Then, I used the following commands to inspect the data:
```{r Data Inspection}
str(fang)
str(snps)
```
I felt that most of the things that could be covered by other functions are included in the str function, but other options for data inspection are names, dim, typeof, length, most of which are covered under str. 

##Data Manipulation

Note: During data manipulation process, I typically used the View command to look at the structure of the data after each step, if possible.

The first step that needs to occur is the removal of the selected genotypes from the overall file to generate two files: One that contains all of the zea mays and one that contains teosinte. I used the dplyr package to filter my data. The first line of code is from lecture, and I used it to determine if my machine had dplyr or not.

```{r filter genotypes}
if (!require("dplyr")) install.packages("dplyr") 
library(dplyr)
maize <- filter(fang, Group == "ZMMIL" | Group == "ZMMLR" | Group == "ZMMMR")
teosinte <- filter(fang, Group == "ZMPBA" | Group == "ZMPIL" | Group == "ZMPJA")
```

The next step is to transpose, so I used:
```{r Transpose}
trans_maize <- t(maize)
trans_teos <- t(teosinte)
```
I saw that the newly transposed objects trans_teos and trans_maize were matrices, and felt the need to convert them to dataframes, which also needs to have the strings as factors turned to false to have manipulatable data. 

```{r Coerce to data frame}
df_trans_maize <- as.data.frame(trans_maize, stringsAsFactors = FALSE)
df_trans_teos <- as.data.frame(trans_teos, stringsAsFactors = FALSE)
```

This gave me transposed data frames. From here, the data needed to be joined together, using the merge command:
```{r Merge files}
maize_merge <- merge(snps, df_trans_maize, by.x = "SNP_ID", by.y = "row.names")
teos_merge <- merge(snps, df_trans_teos, by.x = "SNP_ID", by.y = "row.names")
```
After the merge, I needed to reorder the columns, so I used:
```{r Reorder columns}
rearr_maize <- maize_merge[,c(1,3,4,2,5:1588)]
rearr_teos <- teos_merge[,c(1,3,4,2,5:990)]
```
After checking the order with View, I moved to the next step, which was substituting either "-" or "?" for NA. I found a way to do this using is.na, but the command acts on a data frame, and replaces the NAs in that data frame, and cannot be assigned to a new one. So, I used the following:
```{r NA substitution}
hyph_maize <- rearr_maize
hyph_maize[is.na(hyph_maize)]<- "-"
ques_maize <- rearr_maize
ques_maize[is.na(ques_maize)] <- "?"
hyph_teos <- rearr_teos
hyph_teos[is.na(hyph_teos)]<- "-"
ques_teos <- rearr_teos
ques_teos[is.na(ques_teos)] <- "?"
```
This generated four files, which needed to be sorted, so I used the arrange command as below:
```{r Sort ascending/descending}
arr_hyphmaize <- arrange(hyph_maize, desc(Position))
arr_hyphteos <- arrange(hyph_teos, desc(Position))
arr_quesmaize <- arrange(ques_maize, Position)
arr_questeos <- arrange(ques_teos, Position)
```
After sorting the four files, I needed to generate 10 files per each of the four in the previous step, for a total of 40 files. First I split the data by the chromosome column using:
```{r Split}
hyphmaize_chr_split <- split(arr_hyphmaize, arr_hyphmaize$Chromosome)
quesmaize_chr_split <- split(arr_quesmaize, arr_quesmaize$Chromosome)
hyphteos_chr_split <- split(arr_hyphteos, arr_hyphteos$Chromosome)
questeos_chr_split <- split(arr_questeos, arr_questeos$Chromosome)
```
Then, I used the function lapply to apply a function to all elements of my list, using a function defined as write.table for each element of the list, and creating files with the prefix of th(teosinte hyphens), tq(teosinte question marks),  mh, and mq (maize equivalents).Then, I used followed the prefix with the name "chromosome", followed by the chromosome number (x):
```{r File generation with lapply}
lapply(names(hyphmaize_chr_split), function(x){write.table(hyphmaize_chr_split[[x]], file = paste("mh_chromosome",x, sep = "\t"))})
lapply(names(quesmaize_chr_split), function(x){write.table(quesmaize_chr_split[[x]], file = paste("mq_chromosome",x, sep = "\t"))})
lapply(names(hyphteos_chr_split), function(x){write.table(hyphteos_chr_split[[x]], file = paste("th_chromosome",x, sep = "\t"))})
lapply(names(questeos_chr_split), function(x){write.table(questeos_chr_split[[x]], file = paste("tq_chromosome",x, sep = "\t"))})
```
After generating these 40 files, I deleted the multiple and unknown position files, and moved the 10 relevant files each to a folder named question_files or hyphenated_files, then further into folders based on genotype. 

For the next part of the assignment, ggplots needed to be generated, and it was suggested to use the reshape2 package, so I ran the following to determine if it was installed or not:
```{r install reshape}
if (!require("reshape2")) install.packages("reshape2")
library(reshape2)
```
Once reshape2 was installed and loaded, I usedthe following to make the data tidier:
```{r melt}
molten_maize <- melt(rearr_maize)
molten_teos <- melt(rearr_teos)
```
Then, in order to generate a plot of the number of SNPs in each chromosome, I created a table of chromosome SNP observation frequencies:
```{r}
maize_table <-as.data.frame(table(molten_maize$Chromosome))
colnames(maize_table)[1] <- "Maize_Chrom_Number"
teos_table <-as.data.frame(table(molten_teos$Chromosome))
colnames(teos_table)[1]<- "Teos_Chrom_Number"
```

Then, I used the following to generate a plot of SNPs per chromosome in each genotype:

```{r Install ggplot2}
if (!require("ggplot2")) install.packages("ggplot2")
library(ggplot2)
```

```{r Generate plots of SNPs}
ggplot(maize_table)+geom_col(aes(x=Maize_Chrom_Number, y=Freq))
ggplot(teos_table)+geom_col(aes(x=Teos_Chrom_Number, y=Freq))
```

The next plot that it was necessary to generate was the plot of the amount of heterozygosity in each species and group. For this analysis, my data was not formatted correctly. I noticed that the group column was no longer present, so I created a new dataset using the following commands:
```{r}
fang2 <- as.data.frame(read.table ("working_files/fang_et_al_genotypes.txt", sep = "\t", header = TRUE))
snp2 <- as.data.frame(read.table ("working_files/snp_position.txt", sep = "\t", header = TRUE))
fang2_alt <- fang2[, -2] #remove second column
fang_molten <- melt(fang2_alt, id= c("Sample_ID", "Group")) #melt data to prepare to join to relevant snp info
colnames(fang_molten)[c(3,4)] <- c("SNP_ID", "Nucleotides") #Rename columns to better identify data
snp_rel <- snp2[, c(1,3,4)] #pull relevant info from SNP file, only columns 1,3,4 necessary for plot
```
I only need the info for the 10 chromosomes, not the multiple or unknown data, so we can code those as NA values. 
```{r}
snp_rel[snp_rel == "unknown"| snp_rel == "multiple"] <- NA
```
I also needed to change the Position and Chromosome columns to numeric data, in order to plot and join correctly. 
```{r}
snp_rel$Chromosome <- as.numeric(as.character(snp_rel$Chromosome))
snp_rel$Position <- as.numeric(as.character(snp_rel$Position))
```
Now that the formatting is complete, we simply need to use the merge function to create a data frame that is capable of forming our plot (this may take some time):
```{r}
SNP_total <- merge(snp_rel, fang_molten, by="SNP_ID")
```
This file now has all relevant info, including group name, to use in creating the last plot. We were instructed in the prompt to create a new column indicating the heterozygosity of the data, so I used the following if/else function to determine if a site was heterozygous or not, or if it had NAs (multiple/unknown). 
```{r}
is.hetero <- function(x) {
  if (x=="C/C" | x=="A/A"|x=="T/T" |x=="G/G") {return("Homo.")}
  else if (x=="?/?") {return(NA)}
  else {return("Hetero.")}
}
```
This function was applied to my newly created data set to make a new column named "Heterozygosity".
```{r}
SNP_total$Heterozygosity <- lapply(SNP_total$Nucleotides, is.hetero)
```
Because the new column is not character data, it cannot yet be plotted, so I coerced it:
```{r}
SNP_total$Heterozygosity <- as.character(SNP_total$Heterozygosity)
```
I only need to plot the maize and teosinte data, so I pulled them out to enter into a new data frame:
```{r}
maize_teos_SNP <- filter(SNP_total, Group == "ZMMIL" | Group == "ZMMLR" | Group == "ZMMMR" | Group == "ZMPBA" | Group == "ZMPIL" | Group == "ZMPJA")
maize_teos_SNP$Group <- as.character(maize_teos_SNP$Group)
```
Now the file is complete, and can be plotted with the following to show how many heterozygous, homozygous, and unknown values are present for each group within species:
```{r}
ggplot(maize_teos_SNP, aes(Heterozygosity, ..count..))+geom_bar(aes(fill=Group), position= "dodge")
```


Finally, there was a plot of my own choice, and I chose to visualize the locations of all SNPs within each of their chromosomes. This plot has an extremely cluttered x-axis, but I couldn't unclutter it without converting that column to numeric data, which ggplot2 apparently cannot use. The smaller values start on the left of the x axis, and increase moving right(as is normal on a numerical axis).
```{r}
ggplot(rearr_teos)+geom_point(aes(x=Position, y=Chromosome))
```
I attempted to fix this plot by changing the size and angle of the x axis labels:
```{r}
ggplot(rearr_teos)+geom_point(aes(x=Position, y=Chromosome))+theme(axis.text.x = element_text(face="plain", size = 8, angle=60))
```
But this plot is also unreadable on the x axis. However, the plot still gives us relative locations within each of the chromosomes where we find each SNP. 


